src/enrichment_tables/file.rs

```diff
@@ -1,13 +1,23 @@
-use std::collections::BTreeMap;
-
 use crate::config::{EnrichmentTableConfig, EnrichmentTableDescription};
 use serde::{Deserialize, Serialize};
-use shared::btreemap;
+use std::collections::{BTreeMap, HashMap};
+use std::hash::Hasher;
 use tracing::trace;
-use vector_core::enrichment::{Condition, Table};
+use vector_core::enrichment::{Condition, IndexHandle, Table};
 
 #[derive(Serialize, Deserialize, Debug, Clone, Default, PartialEq)]
-struct FileConfig;
+struct FileConfig {
+    filename: String,
+    encoding: String,
+    #[serde(default = "crate::serde::default_true")]
+    include_headers: bool,
+    #[serde(default = "default_delimiter")]
+    delimiter: char,
+}
```

Comment:
The structure of this type seems off to me.

I'd expect "data source" and "data encoding" to be two different types of data that need to be combined to form a valid enrichment table config.

That is to say, I'd expect `encoding` to be its own type, not part of the `File` source type.

Something like:

```rust
mod source {
	struct File {
		path: PathBuf,
	}
	struct Url {
		url: Url,
	}
	// etc...
}

mod encoding {
	struct Csv {
		include_headers: bool,
		delimiter: String,
	}
	struct Json;
	struct Toml;
	// etc...
}
```

And then the final Vector config would have you define both the table source, and its encoding (and later, optionally decryption options):

```toml
[enrichment_tables.my_csv]
  name = "mycsv"
  type = "file"
  file.path = "./my.csv"
  format = "csv"
  csv.include_headers = true
  csv.delimiter = ";"
```

Reply:
Have updated it so the config now looks like:

```toml
[enrichment_tables.file]
type = "file"

source.type = "file"
source.path = "../MOCK_DATA.csv"

encoding.type = "csv"
encoding.delimiter = ','
encoding.include_headers = true
```

Which I now realise is just duplicating the `file` type.  I'll remove that bit. If we want a different source type, we'd probably create a new enrichment table type. 

Or possibly, this would be a `load` enrichment table. ie.. the data is loaded once. Then source could be file or http or whatever. Another type of enrichment table would be `streaming`.

That's going down to design which we don't need now, so I'll just remove the `source.type`..

Reply:
It is now:

```toml
[enrichment_tables.file]
type = "file"
path = "../MOCK_DATA.csv"
encoding.type = "csv"
encoding.delimiter = ','
encoding.include_headers = true
```

Reply:
Per https://github.com/timberio/vector/issues/8688 I think we should restructure this a bit to be:

```
type = "file"
file.path = "../MOCK_DATA.csv"
encoding.type = "csv"
encoding.csv.delimiter = ','
encoding.csv.include_headers = true
```

This is the config pattern we introduced with the `exec` source and have been continuing to expand the prevalence of. I think the advantage is that it is clear which options apply when.

Reply:
However, I also think that `csv` encoding type is probably only relevant for `type = "file"` so maybe it should even be:

```toml
type = "file"
file.path = "../MOCK_DATA.csv"
file.encoding.type = "csv"
file.encoding.csv.delimiter = ','
file.encoding.csv.include_headers = true
```

?

Reply:
cc/ @binarylogic since he usually has config modeling opinons.

Reply:
I agree with @jszwedko's [last comment](https://github.com/timberio/vector/pull/8730#discussion_r693073723). It has been chosen!

Reply:
Ok, it is:

```toml
[enrichment_tables.file]
type = "file"
file.path = "../MOCK_DATA.csv"
file.encoding.type = "csv"
file.encoding.delimiter = ','
file.encoding.include_headers = true
```